### PR TITLE
Add NULL record

### PR DESCRIPTION
**Pull request by miekg**
Sorely missing from this library. Add it.

### CODE REVIEW

**Comment by tmthrgd** [comment]
@miekg Can this wait until #833? That changes the signature of `len` and it would be much easier to rebase this PR rather than #833.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/dns] WIP: Add NULL recor..." ]
>@miekg Can this wait until #833? That changes the signature of `len` and it would be much easier to rebase this PR rather than #833.

yeah, this can wait.

**Comment by miekg** [comment]
rebased and regenerated. Only needs a little test. Hard to validate if the implementation is interopereable because there is no presentation format.

**Comment by tmthrgd** [comment]
`go install` + `go generate`. You're not picking up the changes from #847.

**Comment by tmthrgd** [comment]
Could you add a test against a fixed record data that was generated by something else?

**Comment by tmthrgd** [comment]
If they’re used by tunneling implementations, maybe one of those could be used to generate the record? It’s not obvious without a presentation format.

**Comment by tmthrgd** [comment]
At the very least, you could add a regression test by using this implementation to generate the record to make sure it doesn’t accidentally change.

File: parse_test.go [hunk 1]
```diff
@@ -398,16 +398,16 @@ func TestQuotes(t *testing.T) {
...
+		`t.example.com. IN TXT "aa"`:                                                         "t.example.com.\t3600\tIN\tTXT\t\"aa\"",
+		`t.example.com. IN TXT "aaa" ;`:                                                      "t.example.com.\t3600\tIN\tTXT\t\"aaa\"",
+		`t.example.com. IN TXT "abc" "DEF"`:                                                  "t.example.com.\t3600\tIN\tTXT\t\"abc\" \"DEF\"",
+		`t.example.com. IN TXT "abc" ( "DEF" )`:                                              "t.example.com.\t3600\tIN\tTXT\t\"abc\" \"DEF\"",
+		`t.example.com. IN TXT aaa ;`:                                                        "t.example.com.\t3600\tIN\tTXT\t\"aaa\"",
+		`t.example.com. IN TXT aaa aaa;`:                                                     "t.example.com.\t3600\tIN\tTXT\t\"aaa\" \"aaa\"",
+		`t.example.com. IN TXT aaa aaa`:                                                      "t.example.com.\t3600\tIN\tTXT\t\"aaa\" \"aaa\"",
+		`t.example.com. IN TXT aaa`:                                                          "t.example.com.\t3600\tIN\tTXT\t\"aaa\"",
```
Comment (tmthrgd) [inline parse_test.go]: @miekg I think you need to update `gofmt`, see https://golang.org/doc/go1.11#gofmt. The original formatting of these lines was correct as of go1.11.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/dns] Add NULL record (#8..." ]
>tmthrgd requested changes on this pull request.
>
>
>
>> -		`t.example.com. IN TXT "abc" "DEF"`:     "t.example.com.\t3600\tIN\tTXT\t\"abc\" \"DEF\"",
>-		`t.example.com. IN TXT "abc" ( "DEF" )`: "t.example.com.\t3600\tIN\tTXT\t\"abc\" \"DEF\"",
>-		`t.example.com. IN TXT aaa ;`:           "t.example.com.\t3600\tIN\tTXT\t\"aaa\"",
>-		`t.example.com. IN TXT aaa aaa;`:        "t.example.com.\t3600\tIN\tTXT\t\"aaa\" \"aaa\"",
>-		`t.example.com. IN TXT aaa aaa`:         "t.example.com.\t3600\tIN\tTXT\t\"aaa\" \"aaa\"",
>-		`t.example.com. IN TXT aaa`:             "t.example.com.\t3600\tIN\tTXT\t\"aaa\"",
>+		`t.example.com. IN TXT ""`:                                                           "t.example.com.\t3600\tIN\tTXT\t\"\"",
>+		`t.example.com. IN TXT "a"`:                                                          "t.example.com.\t3600\tIN\tTXT\t\"a\"",
>+		`t.example.com. IN TXT "aa"`:                                                         "t.example.com.\t3600\tIN\tTXT\t\"aa\"",
>+		`t.example.com. IN TXT "aaa" ;`:                                                      "t.example.com.\t3600\tIN\tTXT\t\"aaa\"",
>+		`t.example.com. IN TXT "abc" "DEF"`:                                                  "t.example.com.\t3600\tIN\tTXT\t\"abc\" \"DEF\"",
>+		`t.example.com. IN TXT "abc" ( "DEF" )`:                                              "t.example.com.\t3600\tIN\tTXT\t\"abc\" \"DEF\"",
>+		`t.example.com. IN TXT aaa ;`:                                                        "t.example.com.\t3600\tIN\tTXT\t\"aaa\"",
>+		`t.example.com. IN TXT aaa aaa;`:                                                     "t.example.com.\t3600\tIN\tTXT\t\"aaa\" \"aaa\"",
>+		`t.example.com. IN TXT aaa aaa`:                                                      "t.example.com.\t3600\tIN\tTXT\t\"aaa\" \"aaa\"",
>+		`t.example.com. IN TXT aaa`:                                                          "t.example.com.\t3600\tIN\tTXT\t\"aaa\"",
>
>@miekg I think you need to update `gofmt`, see https://golang.org/doc/go1.11#gofmt. The original formatting of these lines was correct as of go1.11.

ah.. I was wondering about this. Must be ages since I go-installed gofmt

**Comment by miekg** [comment]
ok, conflict resolved and rebased (manually edited zmsg.go, because it can't 'go install' the right version of the go library to make go generate work). PTAL

**Comment by miekg** [comment]
@tmthrgd I think this is good to go

File: parse_test.go [hunk 2]
```diff
@@ -1567,3 +1567,14 @@ func TestBad(t *testing.T) {
 		}
 	}
 }
+
+func TestNULLRecord(t *testing.T) {
+	// packet captured from iodine
+	packet := `8116840000010001000000000569627a6c700474657374046d69656b026e6c00000a0001c00c000a0001000000000005497f000001`
+	data, _ := hex.DecodeString(packet)
+	msg := new(Msg)
+	msg.Unpack(data)
```
Comment (tmthrgd) [inline parse_test.go]: This error should be checked.

**Pull request merged**